src/Microsoft.OData.Client/DeepInsertSaveResult.cs

```diff
@@ -225,19 +268,259 @@ protected override ODataRequestMessageWrapper CreateRequestMessage(string method
+
+                ODataMaterializerContext materializerContext = new ODataMaterializerContext(responseInfo, base.MaterializerCache);
+                entry = ODataReaderEntityMaterializer.ParseSingleEntityPayload(responseMessageWrapper, responseInfo, entityDescriptor.Entity.GetType(), materializerContext);
+
+                this.headers = headers;
+                this.topLevelEntry = entry;
+            }
+            catch (DataServiceRequestException)
+            {
+                throw;
+            }
+            catch (InvalidOperationException ex)
+            {
+                HeaderCollection headers = new HeaderCollection(this.batchResponseMessage);
+                int statusCode = this.batchResponseMessage == null ? (int)HttpStatusCode.InternalServerError : (int)this.batchResponseMessage.StatusCode;
+                DataServiceResponse dataServiceResponse = new DataServiceResponse(headers, statusCode, Enumerable.Empty<OperationResponse>(), this.IsBatchRequest);
+                throw new DataServiceRequestException(Strings.DataServiceException_GeneralError, ex, dataServiceResponse);
+            }
+
+            HandleDeepInsertResponseInternal(entry, true, entityDescriptor, null);
+
+            return response;
+        }
+
+        private void HandleDeepInsertResponseInternal(MaterializerEntry entry, bool isTopLevelDescriptor, Descriptor descriptor, OperationResponse parentOperationResponse)
+        {
+            OperationResponse response = this.CreateOperationResponse(entry, isTopLevelDescriptor, descriptor, parentOperationResponse);
+
+            List<Descriptor> relatedDescriptors = this.BulkUpdateGraph.GetRelatedDescriptors(descriptor);
+
+            foreach (IMaterializerState nestedItem in entry?.NestedItems)
+            {
+                if (nestedItem is MaterializerNestedEntry materializerNestedEntry)
+                {
+                    foreach (IMaterializerState nestedEntries in materializerNestedEntry.NestedItems)
+                    {
+                        if (nestedEntries is MaterializerFeed feed)
+                        {
+                            for (int i = 0; i < feed.Items.Count; i++)
+                            {
+                                // Ensures the count of the descriptors matches the feed.Items count.
+                                if (relatedDescriptors.Count < i + 1)
+                                {
+                                    break;
+                                }
+
+                                HandleDeepInsertResponseInternal(feed.Items[i] as MaterializerEntry, false, relatedDescriptors[i], response);
+                            }
+                        }
+                        else if (nestedEntries is MaterializerEntry nestedEntry)
+                        {
+                            // In single value navigation property, the related descriptors count must be 1.
+                            if (relatedDescriptors.Count == 0)

```

Comment:
Can we find ourselves here and at the same time find that the `relatedDescriptors` collection has 0 items?

Reply:
If the descriptors have been detached